core/main/src/broker/endpoint_broker.rs

```diff
@@ -0,0 +1,598 @@
+    fn get_broker_with_session(
+        endpoint: RuleEndpoint,
+        callback: BrokerCallback,
+        _: Option<AccountSession>,
+    ) -> Self
+    where
+        Self: Sized,
+    {
+        Self::get_broker(endpoint, callback)
+    }
+
+    fn get_broker(endpoint: RuleEndpoint, callback: BrokerCallback) -> Self;
+    fn get_sender(&self) -> BrokerSender;
+
+    fn prepare_request(&self, rpc_request: &BrokerRequest) -> Result<Vec<String>, RippleError> {
+        let response = Self::update_request(rpc_request)?;
+        Ok(vec![response])
+    }
+
+    /// Adds BrokerContext to a given request used by the Broker Implementations
+    /// just before sending the data through the protocol
+    fn update_request(rpc_request: &BrokerRequest) -> Result<String, RippleError> {
+        let v = Self::apply_request_rule(rpc_request)?;
+        let id = rpc_request.rpc.ctx.call_id;
+        let method = rpc_request.rule.alias.clone();
+        if let Value::Null = v {
+            Ok(json!({
+                "jsonrpc": "2.0",
+                "id": id,
+                "method": method
+            })
+            .to_string())
+        } else {
+            Ok(json!({
+                "jsonrpc": "2.0",
+                "id": id,
+                "method": method,
+                "params": v
+            })
+            .to_string())
+        }
+    }
+
+    /// Generic method which takes the given parameters from RPC request and adds rules using rule engine
+    fn apply_request_rule(rpc_request: &BrokerRequest) -> Result<Value, RippleError> {
+        if let Ok(mut params) = serde_json::from_str::<Vec<Value>>(&rpc_request.rpc.params_json) {
+            if params.len() > 1 {
+                if let Some(last) = params.pop() {
+                    if let Some(filter) = rpc_request
+                        .rule
+                        .transform
+                        .get_filter(super::rules_engine::RuleTransformType::Request)
+                    {
+                        return jq_compile(
+                            last,
+                            &filter,
+                            format!("{}_request", rpc_request.rpc.ctx.method),
+                        );
+                    }
+                    return Ok(serde_json::to_value(&last).unwrap());
+                }
+            } else {
+                return Ok(Value::Null);
+            }
+        }
+        Err(RippleError::ParseError)
+    }
+
+    /// Default handler method for the broker to remove the context and send it back to the
+    /// client for consumption
+    fn handle_response(result: &[u8], callback: BrokerCallback) {
+        let mut final_result = Err(RippleError::ParseError);
+        if let Ok(data) = serde_json::from_slice::<JsonRpcApiResponse>(result) {
+            final_result = Ok(BrokerOutput { data });
+        }
+        if let Ok(output) = final_result {
+            tokio::spawn(async move { callback.sender.send(output).await });
```

Comment:
Does this need to spawn a thread?

Reply:
Yes to release the queue for the next response